codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/protocol/ServerProtocol.kt

```diff
@@ -338,3 +408,24 @@ class ServerRequestBeforeBoxingDeserializedMemberConvertToMaybeConstrainedJsonPa
             else -> emptySection
         }
 }
+
+/**
+ * A customization to, just before we box a recursive member that we've deserialized from CBOR into `T` held in a
+ * variable binding `v`, convert it into `MaybeConstrained` if the target shape can reach a constrained shape.
+ */
+class ServerRequestBeforeBoxingDeserializedMemberConvertToMaybeConstrainedCborParserCustomization(val codegenContext: ServerCodegenContext) :
+    CborParserCustomization() {
+    override fun section(section: CborParserSection): Writable =
+        when (section) {
+            is CborParserSection.BeforeBoxingDeserializedMember ->
+                writable {
+                    // We're only interested in _structure_ member shapes that can reach constrained shapes.
+                    if (
+                        codegenContext.model.expectShape(section.shape.container) is StructureShape &&
```

Comment:
Hmm being constrained is a property of a Smithy shape, not of a Rust symbol. We're adding `.into()` here to convert to `MaybeConstrained`, _not_ to get the underlying type.

Reply:
Yes, a Smithy shape is constrained, but somewhere in the codebase, we've already generated a Rust symbol that represents this fact. `MaybeConstrained` is a Rust type/symbol, not directly related to Smithy. My point is, we have 108 instances where we check if a Smithy shape is constrained. Instead, why don't we query the symbol provider or the symbol itself to determine if it supports the `MaybeConstrained` Rust type and if we need to issue a `v.into()` to convert to `MaybeConstrained`, rather than checking that a Shape's target can reach a constrained shape, and assuming we've generated a `MaybeConstrained` for that shape?